src/miner.cpp

```diff
@@ -645,42 +663,50 @@ void BitcoinMiner(CWallet* pwallet, bool fProofOfStake)
     bool fLastLoopOrphan = false;
     while (fGenerateBitcoins || fProofOfStake) {
         if (fProofOfStake) {
+            if (chainActive.Tip()->nHeight < Params().LAST_POW_BLOCK()) {
+                //  The last PoW block hasn't even been mined yet.
+                MilliSleep(Params().TargetSpacing() * 1000);       // sleep a block
+                continue;
+            }
+
             //control the amount of times the client will check for mintable coins
             if ((GetTime() - nMintableLastCheck > 5 * 60)) // 5 minute check time
             {
                 nMintableLastCheck = GetTime();
                 fMintableCoins = pwallet->MintableCoins();
             }
 
-            if (chainActive.Tip()->nHeight < Params().LAST_POW_BLOCK()) {
-                MilliSleep(5000);
-                continue;
-            }
-
-            while (vNodes.empty() || pwallet->IsLocked() || !fMintableCoins || (pwallet->GetBalance() > 0 && nReserveBalance >= pwallet->GetBalance()) || !masternodeSync.IsSynced()) {
+            while (vNodes.empty() || pwallet->IsLocked() || !fMintableCoins ||
+                   (pwallet->GetBalance() > 0 && nReserveBalance >= pwallet->GetBalance()) || !masternodeSync.IsSynced()) {
                 nLastCoinStakeSearchInterval = 0;
+                MilliSleep(5000);
```

Comment:
Might as well remove this sleep instead of moving it here, no?

Reply:
The miner thread would then go into a very tight spiral in that while loop.  The better question is "why sleep 5 seconds when the MintableCoins is only called every minute".  Well, it makes the reaction time every 5 seconds.    For example, if you were to watch that area of code and then lock your wallet.  It's going to sit there spinning, checking if it's unlocked every 5 seconds.  Take that sleep out... and it's going to go into an uncontrolled spiral until you unlock your wallet.

Reply:
Why move it at all then?

Reply:
Let's look at the "before"

```
if (some stuff isn't ready)
    if (a minute has gone by since the last computationally expensive check)
        (run the computationally expensive check)   
    wait 5 seconds
    (go back and check if all the stuff is ready yet)
```
So... what happens if a minute has gone by, and there still aren't any mintable coins?  you wait 5 seconds and check everything else, and a minute later you check for mintable coins again.

But now... what happens if it was 1 second before there were mintable coins?  You run the check, then you wait 5 seconds, and then you loop.... and you already checked, so you wait a minute before checking again.   So now it's been a minute before you actually check again.

By moving that sleep; so you're not sleeping after the check... you do the requisite 5 second wait; then you check for mintable coins; that were ready 4 seconds ago... and you're now off to the races only 4 seconds after they were ready.  Where the original code; you're not off to the races until 59 seconds after the mint was ready.

So boil that on up to what it means?  Well, if block creation was lagging the desired target spacing; then blocks will start being written...less than 1 minute in between.  Presumably, it was the confirmation from the last block that took you from fMintableCoins == false to fMintableCoins == true.   If you're out of that loop 4 seconds after the last block; you have a shot at submitting the next block.  If you're out of that loop 59 seconds after the last block; you probably don't have a shot of submitting the next block.

ergo:  "efficiency". :)